src/snowflake/connector/s3_storage_client.py

```diff
@@ -87,84 +92,145 @@ def __init__(
+
+        Does not support sorting by values in case the keys are the same, don't send
+        in duplicate keys, but this is not possible with a dictionary anyways.
+        """
         res = []
+        low_key_dict = {k.lower(): v for k, v in headers.items()}
+        sorted_headers = sorted(low_key_dict.keys())
+        _res = [(k, low_key_dict[k]) for k in sorted_headers]
 
-        for i in range(len(_res)):
-            k, v = _res[i]
+        for k, v in _res:
             # if value is a list, convert to string delimited by comma
             if isinstance(v, list):
                 v = ",".join(v)
             # if multiline header, replace withs space
             k = k.replace("\n", " ")
-            res.append(k.rstrip() + ":" + v.lstrip())
+            res.append(k.strip() + ":" + RE_MULTIPLE_SPACES.sub(" ", v.strip()))
 
         ans = "\n".join(res)
         if ans:
-            ans = ans + "\n"
+            ans += "\n"
 
-        return ans
+        return ans, ";".join(sorted_headers)
 
     @staticmethod
-    def _construct_string_to_sign(
+    def _construct_canonical_request_and_signed_headers(
         verb: str,
-        canonicalized_element: str,
-        canonicalized_headers: str,
-        amzdate: str,
-        content_md5: str = "",
-        content_type: str = "",
-    ) -> bytes:
-        res = verb + "\n" + content_md5 + "\n" + content_type + "\n"
-        res += amzdate + "\n" + canonicalized_headers + canonicalized_element
-        return res.encode("UTF-8")
+        canonical_uri_parameter: str,
+        query_parts: Dict[str, str],
+        canonical_headers: Optional[Dict[str, Union[str, List[str]]]] = None,
+        payload_hash: str = "",
+    ) -> Tuple[str, str]:
+        """Build canonical request and also return signed headers.
+
+        Note: this doesn't support sorting by values in case the same key is given
+         more than once, but doing this is also not possible with a dictionary.
+        """
+        canonical_query_string = "&".join(
+            "=".join([k, v]) for k, v in sorted(query_parts.items(), key=itemgetter(0))
+        )
+        (
+            canonical_headers,
+            signed_headers,
+        ) = SnowflakeS3RestClient._construct_canonicalized_and_signed_headers(
+            canonical_headers
+        )
+
+        return (
+            "\n".join(
+                [
+                    verb,
+                    canonical_uri_parameter or "/",
+                    canonical_query_string,
+                    canonical_headers,
+                    signed_headers,
+                    payload_hash,
+                ]
+            ),
+            signed_headers,
+        )
 
     @staticmethod
-    def _has_expired_token(response: requests.Response) -> bool:
+    def _construct_string_to_sign(
+        region_name: str,
+        service_name: str,
+        amzdate: str,
+        canonical_request_hash: bytes,
+    ) -> Tuple[str, str]:
```

Comment:
This does part of step 3 from my explanation https://github.com/snowflakedb/snowflake-connector-python/pull/846#issuecomment-914643727 the result of this then gets hashed.